src/miner_lora.erl

```diff
@@ -0,0 +1,412 @@
+    {ok, Socket} = gen_udp:open(UDPPort, [binary, {active, 100}, {ip, UDPIP}]),
+    {ok, #state{socket=Socket,
+                sig_fun = maps:get(sig_fun, Args),
+                pubkey_bin = blockchain_swarm:pubkey_bin()}}.
+
+handle_call({send, Payload, When, Freq, DataRate, Power}, From, #state{socket=Socket,
+                                                                       gateways=Gateways,
+                                                                       packet_timers=Timers}=State) ->
+    Token = mk_token(Timers),
+    %% TODO we should check this for regulatory compliance
+    BinJSX = jsx:encode(
+        #{<<"txpk">> => #{
+            %% TODO don't use inverse polarity for longfi?
+            <<"ipol">> => true,
+            <<"imme">> => When == immediate,
+            <<"powe">> => trunc(Power),
+            %% TODO gps time?
+            <<"tmst">> => When,
+            <<"freq">> => Freq,
+            <<"modu">> => <<"LORA">>,
+            <<"datr">> => list_to_binary(DataRate),
+            <<"codr">> => <<"4/5">>,
+            <<"size">> => byte_size(Payload),
+            <<"rfch">> => 0,
+            <<"data">> => base64:encode(Payload)
+        }
+    }),
+    #gateway{ip=IP, port=Port} = select_gateway(Gateways),
+    Packet = <<?PROTOCOL_2:8/integer-unsigned, Token/binary, ?PULL_RESP:8/integer-unsigned, BinJSX/binary>>,
+    ok = gen_udp:send(Socket, IP, Port, Packet),
+    %% TODO a better timeout would be good here
+    Ref = erlang:send_after(10000, self(), {tx_timeout, Token}),
+    {noreply, State#state{packet_timers=maps:put(Token, {send, Ref, From}, Timers)}};
+handle_call(_Msg, _From, State) ->
+    lager:warning("rcvd unknown call msg: ~p", [_Msg]),
+    {reply, ok, State}.
+
+handle_cast(_Msg, State) ->
+    lager:warning("rcvd unknown cast msg: ~p", [_Msg]),
+    {noreply, State}.
+
+handle_info({tx_timeout, Token}, #state{packet_timers=Timers}=State) ->
+    case maps:find(Token, Timers) of
+        {ok, {send, _Ref, From}} ->
+            gen_server:reply(From, {error, timeout});
+        error ->
+            ok
+    end,
+    {noreply, State#state{packet_timers=maps:remove(Token, Timers)}};
+handle_info({udp, Socket, IP, Port, Packet}, #state{socket=Socket}=State) ->
+    State2 = handle_udp_packet(Packet, IP, Port, State),
+    {noreply, State2};
+handle_info({udp_passive, Socket}, #state{socket=Socket}=State) ->
+    inet:setopts(Socket, [{active, 100}]),
+    {noreply, State};
+handle_info(_Msg, State) ->
+    lager:warning("rcvd unknown info msg: ~p", [_Msg]),
+    {noreply, State}.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+terminate(_Reason, #state{socket=Socket}) ->
+    gen_udp:close(Socket),
+    ok.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+
+-spec mk_token(map()) -> binary().
+mk_token(Timers) ->
```

Comment:
how long do these live?  are they important for protocol reasons?  I suppose it's not super likely that we'll issue two in a row with the same token on a restart, but it worries me a little.

also remember that `strong_rand_bytes` takes a lock in the crypto runtime, so if performance will be at all sensitive, you should use a cheaper form of randomness.